PUBDEV-7860: Add Thin Plate Regression Smoothers to GAM

wendycwong opened the pull request:
This PR completes the work in JIRA: https://h2oai.atlassian.net/browse/PUBDEV-7860.  I have attached a document in the JIRA which is helpful to understand my implementation.

I have completed the following:
1. added thin plate regression smoothers to GAM which involves building the distance measure and the polynomial basis;
2. fixed the model building pipeline to accept both cubic spline smoothers and thin plate regression smoothers
3. completely fix the model building and scoring pipeline
4. add TP to GAM Mojo models;
5. add TP to Python/R clients.  Thanks to @sebhrusen, user can specify gam_columns in as ["c1", ["c2", "c3"]] or [["c1"], ["c2","c3"]] both in both R and Python.
6. fix gridsearch.  Again, thanks 2 @sebhrusen, users are able to specify gam_columns using both notations
7. Added Java tests to completely test out the TP smoothers generation process including distance calculation, penalty matrices generation, polynomial basis generation, TP/CS smoothers with validation dataset, TP/CS smoothers with cross-validation
8. added R/Python unit tests to make sure gam_columns can be specified in both mode, gridsearch and mojo tests.
9. added test to compare our results to R results.  Our gam performs better.  Yes!

There are 18 test files.

wendycwong commented:
[Uploading GamThinPlateRegressionH2O.pdf…]()

wendycwong commented:
[GamThinPlateRegressionH2O.pdf](https://github.com/h2oai/h2o-3/files/5985129/GamThinPlateRegressionH2O.pdf)

maurever commented on h2o-algos/src/main/java/hex/gam/GAM.java:
```diff
@@ -212,59 +233,34 @@ private void validateGamParameters() {
+    sortGAMParameters(_parms, _cubicSplineNum, _thinPlateSmoothersWithKnotsNum); // move cubic spline to the front and thin plate to the back
+    checkThinPlateParams();
     if (_parms._saveZMatrix && ((_train.numCols() - 1 + _parms._num_knots.length) < 2))
       error("_saveZMatrix", "can only be enabled if we number of predictors plus" +
```
?
```suggestion
      error("_saveZMatrix", "can only be enabled if the number of predictors plus" +
```

wendycwong replied:
Thanks.

maurever commented on h2o-algos/src/main/java/hex/gam/GAM.java:
```diff
@@ -284,15 +280,98 @@ private void validateGamParameters() {
     }
   }
+  
+  // verify and check thin plate regression smoothers specific parameters
```
Please, if it is possible, use a valid Java string for documentation.

```
/**
*
**/
```

maurever commented on h2o-algos/src/main/java/hex/gam/GAM.java:
```diff
@@ -284,15 +280,98 @@ private void validateGamParameters() {
+            defaultRows = Math.max(defaultRows, _parms._M[tpCount] + 2);
+            tpCount++;
+          }
+          _parms._num_knots[index] = eligibleRows < defaultRows ? (int) eligibleRows : defaultRows;
```
Are you sure the `eliglibleRows` can't be larger than `int`?

wendycwong replied:
The algorithm will be very slow if num_knot is too big.  Hence, I am restricting it to be integer.

koniecsveta replied:
wouldn't then be better to define `eliglibleRows` directly as integer?

wendycwong replied:
It depends.  If the number of rows in the dataframe is larger than an integer, there can be a problem doing

int eligibleRows = (int) frame.nrows().

koniecsveta replied:
i see, then it's ok

maurever commented on h2o-algos/src/main/java/hex/gam/GAM.java:
```diff
@@ -310,115 +389,259 @@ protected int nModelsInParallel(int folds) {
-    String[][] _gamColNamesCenter;  // gamColNames after de-centering is performed.
+    double[][][] _zTranspose;         // store transpose(Z) matrices for CS and TP smoothers
+    double[][][] _zTransposeCS;       // store transpose(zCS) for thin plate smoother to remove optimization constraint
+    double[][][] _penalty_mat_center; // store centered penalty matrices of all smoothers
```
Is there some reason for using an underscore instead of camel case?

wendycwong replied:
I changed it to camel case.

maurever commented on h2o-algos/src/main/java/hex/gam/GAMModel.java:
```diff
@@ -252,21 +257,31 @@ private void fillUpCoeffs(double[] coeffValues, double[] coeffValuesStand, TwoDi
+    public int[] _num_knots; // array storing number of knots per smoother
+    public int[] _num_knots_sorted;
+    public int[] _num_knots_tp; // store num_knots for thin plate regression
+    public String[] _knot_ids;  // store frame keys that contain knots location for each smoother in gam_X;
```
Again, underscore.

wendycwong replied:
These parameters are used by R/Python clients which use underscore.

maurever replied:
:+1:

maurever commented on h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateDistanceWithKnots.java:
```diff
@@ -0,0 +1,105 @@
+import static org.apache.commons.math3.util.CombinatoricsUtils.factorial;
+import static water.util.ArrayUtils.transpose;
+
+// Implementation details of this class can be found in GamThinPlateRegressionH2O.doc attached to this 
```
:+1: for the link. But please, use java docstring for classes.

wendycwong replied:
Done.

maurever commented on h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GenerateGamMatrixOneColumn.java:
```diff
@@ -84,27 +84,28 @@ public void postGlobal() {  // scale the _penalty function according to R
-    double twoOmagSq = 2.0 / ArrayUtils.innerProduct(_u, _u);
-    for (int rowIndex = 0; rowIndex < _numKnots; rowIndex++) {  // form Z matrix transpose here
-      for (int colIndex = 0; colIndex < _numKnots; colIndex++) {  // skip the first column
+  public static double[][] generateZtransp(Frame gamX, int numKnots) {
```
```suggestion
  public static double[][] generateZTransp(Frame gamX, int numKnots) {
```

wendycwong replied:
Good catch.

maurever commented on h2o-algos/src/main/java/hex/schemas/GAMV3.java:
```diff
@@ -228,20 +231,35 @@
     @API(help="Save keys of model matrix", level = Level.secondary, direction = Direction.INPUT)
     public boolean keep_gam_cols; // if true will save keys storing GAM columns
+
+    @API(help="standardize TP predictor columns", level = Level.secondary, direction = Direction.INPUT)
```
Please, explain what the TP shortcut is in the documentation (if it is not too complicated).

wendycwong replied:
No problem.

wendycwong replied:
Done

maurever commented on h2o-r/tests/testdir_javapredict/runit_PUBDEV_7860_binomial_gam_TP_CS_MOJO.R:
```diff
@@ -0,0 +1,34 @@
+  h2o.downloadCSV(twoFrames$h2oPredict, sprintf("%s/h2oPred.csv", modelAndDir$dirName))
+  h2o.downloadCSV(twoFrames$mojoPredict, sprintf("%s/mojoOut.csv", modelAndDir$dirname))
+  twoFrames$h2oPredict[,1] <- h2o.asfactor(twoFrames$h2oPredict[,1])
+  compareFrames(twoFrames$h2oPredict,twoFrames$mojoPredict, prob=1, tolerance = 1e-6)
```
Why is tolerance more relaxed (1e-6) than for example in `runit_PUBDEV_7860_cartesian_gridsearch_gaussian_dual_model.R` (1e-10)? I am just curious.:)

wendycwong replied:
I copied the test from runit_PUBDEV_7185_quasibinomial_gam_MOJO.R.  There, 1e-6 is used.

maurever replied:
Ok.

maurever reviewed (commented):
@wendy, thank you for this super difficult algorithm improvement. 

I left just a few comments, which could help to improve readability and documentation.

wendycwong commented on h2o-algos/src/main/java/hex/gam/GAM.java:
```diff
@@ -103,32 +119,58 @@ public void cv_computeAndSetOptimalParameters(ModelBuilder[] cvModelBuilders) {
+   * gam_column and then extract the quantiles of that sorted gam_columns.  Here, instead of taking the value for
+   * one gam column, we take the whole row with all the predictors for that smoother.
+   *
+   * @return double[][] array containing the knots specified by users
```
Good catch.  Thanks!

koniecsveta commented on h2o-algos/src/main/java/hex/schemas/GAMV3.java:
```diff
@@ -228,20 +231,35 @@
 
-    @API(help = "Predictor column names for gam", required = true, level = Level.critical, gridable = true)
-    public String[] gam_columns;
+/*    @API(help = "Number of eigenvectors to use in SVD for thin plate regression.  If specified, must be of the same" +
```
commented out code (also L76)

wendycwong replied:
Removed.

koniecsveta commented on h2o-algos/src/main/java/hex/gam/GAM.java:
```diff
@@ -212,61 +233,36 @@ private void validateGamParameters() {
+    } else { // check and make sure gam_columns column types are legal
+      if (_parms._bs == null)
+        setDefaultBSType(_parms);
+      if ((_parms._bs != null) && (_parms._gam_columns.length != _parms._bs.length))  // check length
```
I personally prefer using curly braces even for one line if statements and rather avoid using redundant braces inside if conditions but leaving this up to you

wendycwong replied:
Got it.  The reason for the parenthesis inside if is because I ran into the problem of not remembering the priority of the various operators.  That is why to be sure, I just add parenthesis.

koniecsveta commented on h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateRegressionUtils.java:
```diff
@@ -0,0 +1,347 @@
+import static hex.gam.GAMModel.GAMParameters;
+import static water.util.ArrayUtils.maxValue;
+
+// This class contains functions that perform different functions for generating the thin plate regression splines
```
could be javadoc

wendycwong replied:
Done.

maurever replied:
@wendycwong, I cant see it is Javadoc, am I missing something?

koniecsveta commented on h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateRegressionUtils.java:
```diff
@@ -0,0 +1,347 @@
+    for (int index = 0; index < listSize; index++) {
+      Integer[] oneBasis = onePolyBasis.get(index);
+      int[] freqTable = generateOrderFreq(oneBasis);  // find polynomial basis order and count
+      ArrayList<ArrayList<Integer>> basisPermuations = new ArrayList<>();
```
`List` interface should be used for variable type declaration, already used on some places but in some not

wendycwong replied:
Done.  Thanks.

koniecsveta commented on h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateRegressionUtils.java:
```diff
@@ -0,0 +1,347 @@
+    if (totDegree == 0) {
+      if (currCombo != null)
+        allCombos.add(currCombo.clone());
+      return;
```
redundant return statement

wendycwong replied:
removed.  Thanks.

koniecsveta commented on h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateRegressionUtils.java:
```diff
@@ -0,0 +1,347 @@
+        allCombos.add(currCombo.clone());
+      return;
+    } else if (totDegree < 0 || index >= degreeCombo.length) { // last added value too big
+      return;
```
redundant if-else branch, condition could be modified and merged with else branch

wendycwong replied:
Done.  Thanks.

koniecsveta commented on h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateRegressionUtils.java:
```diff
@@ -0,0 +1,347 @@
+    return knots;
+  }
+  
+  // this class performa scaling on TP penalty matrices that is done in R.
```
could be javadoc

wendycwong replied:
Done.

koniecsveta commented on h2o-core/src/main/java/water/util/ArrayUtils.java:
```diff
@@ -874,6 +885,12 @@ public static long maxValue(long[] from) {
   }
+  public static int maxValue(Integer[] from) {
+    int result = from[0];
+    for (int i = 1; i<from.length; ++i)
```
in this case both, `++i` or `i++` expressions work the same, however, the convention is `i++`. Personally, I would use java 8 Stream API for finding maximum here, since it's more bulletproof.

wendycwong replied:
I switched to stream.  Thanks!

maurever commented:
@wendy, looks like some tests are failing, are they related? Also, INFO Check should pass.

wendycwong commented:
Hi @maurever 

The info check complaint is about using both lower case and upper case in a parameter name.  I fixed that.

Here are the test failures:
R3.5 Medium-large / r_suite.runit_pubdev_3847_gbm_random_attack_large.R : fat finger, added o after locate.  I fixed that and generated a PR here: https://github.com/h2oai/h2o-3/pull/5359
Py3.7 Small / r_suite.pyunit_xgboost_pojo.py : pojo comparison failed.  Not related.
R3.5 Demos Medium-large / r_suite.rdemo.random.data.svd.large.R : test time out.

koniecsveta commented on h2o-py/tests/testdir_algos/xgboost/pyunit_xgboost_pojo.py:
```diff
@@ -25,7 +25,7 @@ def compare_preds(train, test, x, y, booster, ntrees, max_depth, max_error):
 
 
-def test_booster(booster, max_error=1e-10):
+def test_booster(booster, max_error=1e-6):
```
is this good enough?

wendycwong replied:
Zuzana:  Talked to michalk about this.  The accuracy change is due to change from float to double or vice versa in XGBoost.  It has nothing to do with my code.  He is okay with that.

koniecsveta commented on h2o-r/h2o-package/R/models.R:
```diff
@@ -275,9 +282,11 @@ NULL
     }
-    e
-  })
+    e })
```
would leave formatting as it was

wendycwong replied:
I reverted it back to as it was

koniecsveta commented on h2o-genmodel/src/main/java/hex/genmodel/algos/gam/GamUtilsThinPlateRegression.java:
```diff
@@ -0,0 +1,53 @@
+  public static double calTPConstantTerm(int m, int d, boolean dEven) {
+    if (dEven)
+      return (Math.pow(-1, m + 1 + d / 2.0) / (Math.pow(2, 2*m - 1) * Math.pow(Math.PI, d / 2.0) * 
+              factorial(m-1)*factorial(m - d / 2)));
```
is this supposed to calculate terms before r_i from 3.1? If yes, why there is factorial(m-1) term?

wendycwong replied:
Zuzana:

Thank you for checking it.  I forgot to upload the latest formula to the JIRA.  Please download the new document from the JIRA.  The correct formula is 
<img width="1107" alt="image" src="https://user-images.githubusercontent.com/7231712/111013841-de4d5180-8355-11eb-8eb9-5c4ffacf4451.png">

koniecsveta replied:
okok thanks!

wendycwong commented:
I have performed a benchmark to see if adding thin plate has changed much in performance.  Here are the results:

<img width="548" alt="image" src="https://user-images.githubusercontent.com/7231712/111013930-456b0600-8356-11eb-90ce-7f9f23b2b694.png">

maurever commented on h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateRegressionUtils.java:
```diff
@@ -0,0 +1,346 @@
+    return polyBasisArr;
+  }
+  
+  // generate knots for TP smoothers.  Sort the first predictor column, take quantiles out of the sorted
```
Same here, it is just cosmetics, however, Javadoc can really help read the documentation.

wendycwong replied:
@maurever  I changed both of them to java docs.  Thanks.

maurever commented on h2o-py/tests/testdir_algos/gam/pyunit_PUBDEV_7860_PUBDEV_7367_random_gridsearch_subspaces_TP_CS_binomial.py:
```diff
@@ -0,0 +1,104 @@
+from h2o.estimators.gam import H2OGeneralizedAdditiveEstimator
+from h2o.grid.grid_search import H2OGridSearch
+
+# I copied Karthik's test to perform gridsearch on TP/CS smoothers with random gridsearch.  
```
Is this comment still relevant?

wendycwong replied:
Yes, this test is taken from Karthik and I just change the parameters and dataset.  I want to give him the credit and that is why the comment is there.

maurever reviewed (approved):
@wendycwong, the PR looks good to me. Thanks for it. 

I have just a few nitpicking comments which can improve documentation. But it is up to you to change it or not.

koniecsveta commented:
LGTM after conflicts are resolved :ok_hand:

maurever reviewed (approved):
LGTM, thanks @wendycwong for this huge contribution.

michalkurka commented:
@wendycwong I did a high-level review and found no issues, LGTM!

Thanks for your effort

ledell commented on h2o-r/h2o-package/R/gam.R:
```diff
@@ -15,7 +15,8 @@
 #' @param training_frame Id of the training data frame.
-#' @param gam_columns Predictor column names for gam
+#' @param gam_columns Arrays of predictor column names for gam for smoothers using single or multiple predictors like
+#'        {{'c1'},{'c2','c3'},{'c4'},...}
```
@wendycwong I know that this text generated from the Java docs using `gen_gam.py`, but I wonder if we can post-process to if from: `{{'c1'},{'c2','c3'},{'c4'},...}` to use R code?  As a user, I might get confused and perhaps not know if it should be specified as:  `c(c("c1"),c("c2", " c3"), c("c4")))` or `list(c("c1"),c("c2", " c3"), c("c4")))`.  Since it's list in R and Python, we could say "list of predictor column names" vs "arrays of predictor column names" (I think we have a precedent of describing things in the API as "lists" even if they are technically arrays in Java, for clarity to the R/Py users).

ledell commented on h2o-r/h2o-package/R/gam.R:
```diff
@@ -116,9 +117,12 @@
-#' @param bs Basis function type for each gam predictors, 0 for cr
-#' @param scale Smoothing parameter for gam predictors
+#' @param knot_ids String arrays storing frame keys of knots.  One for each gam column set specified in gam_columns
+#' @param standardize_tp_gam_cols \code{Logical}. standardize tp (thin plate) predictor columns Defaults to FALSE.
```
Minor updates: `Standardize tp (thin plate) predictor columns.  Defaults to FALSE.`

ledell commented on h2o-r/h2o-package/R/gam.R:
```diff
@@ -116,9 +117,12 @@
-#' @param knot_ids String arrays storing frame keys of knots.  One for each gam column specified in gam_columns
-#' @param bs Basis function type for each gam predictors, 0 for cr
-#' @param scale Smoothing parameter for gam predictors
+#' @param knot_ids String arrays storing frame keys of knots.  One for each gam column set specified in gam_columns
```
Some descriptions have periods and some are missing (I think it's more common to have periods, so I would switch it to that format).

ledell commented on h2o-r/h2o-package/R/gam.R:
```diff
@@ -116,9 +117,12 @@
+#' @param standardize_tp_gam_cols \code{Logical}. standardize tp (thin plate) predictor columns Defaults to FALSE.
+#' @param scale_tp_penalty_mat \code{Logical}. Scale penalty matrix for tp (thin plate) smoothers as in R Defaults to FALSE.
+#' @param bs Basis function type for each gam predictors, 0 for cr, 1 for thin plate regression with knots, 2 for thin
+#'        plate regression with SVD.  If specified, must be the same size as gam_columns
```
unintentional line break?

ledell reviewed (commented):
Hi @wendycwong I have some minor comments/suggestions (that you could add to a follow-up PR if you want)?
